acorn/index.d.ts

```diff
@@ -26,43 +105,151 @@ declare namespace acorn {
+        _while: TokenType;
+        _with: TokenType;
+        _new: TokenType;
+        _this: TokenType;
+        _super: TokenType;
+        _class: TokenType;
+        _extends: TokenType;
+        _export: TokenType;
+        _import: TokenType;
+        _null: TokenType;
+        _true: TokenType;
+        _false: TokenType;
+        _in: TokenType;
+        _instanceof: TokenType;
+        _typeof: TokenType;
+        _void: TokenType;
+        _delete: TokenType;
     }
 
+    class TokContext {
+        constructor(token: string, isExpr: boolean, preserveSpace: boolean, override: (p: Parser) => void);
+    }
+
+    const tokContexts: {
+        b_stat: TokContext;
+        b_expr: TokContext;
+        b_tmpl: TokContext;
+        p_stat: TokContext;
+        p_expr: TokContext;
+        q_tmpl: TokContext;
+        f_expr: TokContext;
+    };
+
+    function isIdentifierStart(code: number, astral?: boolean): boolean;
+
+    function isIdentifierChar(code: number, astral?: boolean): boolean;
+
     interface AbstractToken {
         start: number;
         end: number;
-        loc: ESTree.SourceLocation;
-        range: [number, number];
+        loc?: SourceLocation;
+        range?: [number, number];
+    }
+
+    interface Comment extends AbstractToken {
+        type: string;
+        value: string;
+    }
+
+    class Token {
+        constructor(p: Parser);
     }
 
     interface Token extends AbstractToken {
         type: TokenType;
         value: any;
     }
 
-    interface Comment extends AbstractToken {
-        type: string;
-        value: string;
+    function isNewLine(code: number): boolean;
+
+    const lineBreak: RegExp;
+
+    const lineBreakG: RegExp;
+
+    const version: string;
+
+    interface IParse {
+        (input: string, options?: Options): ESTree.Program;
     }
 
-    interface Options {
-        ecmaVersion?: number;
-        sourceType?: string;
-        onInsertedSemicolon?: (lastTokEnd: number, lastTokEndLoc?: ESTree.Position) => any;
-        onTrailingComma?: (lastTokEnd: number, lastTokEndLoc?: ESTree.Position) => any;
-        allowReserved?: boolean;
-        allowReturnOutsideFunction?: boolean;
-        allowImportExportEverywhere?: boolean;
-        allowHashBang?: boolean;
-        locations?: boolean;
-        onToken?: ((token: Token) => any) | Token[];
-        onComment?: ((isBlock: boolean, text: string, start: number, end: number, startLoc?: ESTree.Position, endLoc?: ESTree.Position) => any) | Comment[];
-        ranges?: boolean;
-        program?: ESTree.Program;
-        sourceFile?: string;
-        directSourceFile?: string;
-        preserveParens?: boolean;
-        plugins?: { [name: string]: Function; };
+    const parse: IParse;
+
+    function parseExpressionAt(input: string, pos: number, options?: Options): ESTree.Expression;
+
+    function tokenizer(): Parser;
+
+    let parse_dammit: IParse;
+    let LooseParser: LooseParser;
+    let pluginsLoose: PlainObject;
```

Comment:
Since we're targetting latest stable TypeScript version, these should be of types `IParse | undefined`, `LooseParser | undefined`, ... to correctly work with `strictNullChecks`.